tests/catch/unit/memory/hipMemsetSync.cc

```diff
@@ -0,0 +1,505 @@
+                            hipMemcpyDeviceToHost));
+      break;
+
+    default:
+      HIP_CHECK(hipMemcpy(hostMem, aPtr, sizeInBytes, hipMemcpyDeviceToHost));
+      break;
+  }
+}
+
+template <typename T>
+static void hostCopy(memSetType memType, T* aPtr, T* hostMem, size_t dataW, size_t dataH,
+                     size_t dataD, size_t& dataPitch) {
+  size_t elementSize = sizeof(T);
+  size_t sizeInBytes = elementSize * dataW * dataH * dataD;
+  hipMemcpy3DParms params{};
+  switch (memType) {
+    case memSetType::hipMemset3D: {
+      params.kind = hipMemcpyHostToHost;
+      params.srcPos = make_hipPos(0, 0, 0);
+      params.dstPos = make_hipPos(0, 0, 0);
+      params.srcPtr = make_hipPitchedPtr(aPtr, dataPitch, dataW, dataH);
+      params.dstPtr = make_hipPitchedPtr(hostMem, dataW, dataW, dataH);
+
+      hipExtent extent;
+      extent.width = dataW;
+      extent.height = dataH;
+      extent.depth = dataD;
+
+      params.extent = extent;
+
+      HIP_CHECK(hipMemcpy3D(&params));
+      break;
+    }
+
+    case memSetType::hipMemset2D:
+      HIP_CHECK(hipMemcpy2D(hostMem, dataW * elementSize, aPtr, dataPitch, dataW, dataH,
+                            hipMemcpyHostToHost));
+      break;
+
+    default:
+      HIP_CHECK(hipMemcpy(hostMem, aPtr, sizeInBytes, hipMemcpyHostToHost));
+      break;
+  }
+}
+
+template <typename T>
+static void devRegisteredCopy(memSetType memType, T* aPtr, T* hostMem, size_t dataW, size_t dataH,
+                              size_t dataD, size_t& dataPitch) {
+  size_t elementSize = sizeof(T);
+
+  switch (memType) {
+    case memSetType::hipMemset3D: {
+      hipMemcpy3DParms params{};
+      params.kind = hipMemcpyHostToHost;
+      params.srcPos = make_hipPos(0, 0, 0);
+      params.dstPos = make_hipPos(0, 0, 0);
+      params.srcPtr = make_hipPitchedPtr(aPtr, dataPitch, dataW, dataH);
+      params.dstPtr = make_hipPitchedPtr(hostMem, dataW, dataW, dataH);
+
+      hipExtent extent;
+      extent.width = dataW;
+      extent.height = dataH;
+      extent.depth = dataD;
+
+      params.extent = extent;
+
+      HIP_CHECK(hipMemcpy3D(&params));
+      break;
+    }
+
+    case memSetType::hipMemset2D:
+      HIP_CHECK(hipMemcpy2D(hostMem, dataW * elementSize, aPtr, dataPitch, dataW, dataH,
+                            hipMemcpyDeviceToHost));
+      break;
+
+    default: {
+      size_t sizeInBytes = elementSize * dataW * dataH * dataD;
+      HIP_CHECK(hipMemcpy(hostMem, aPtr, sizeInBytes, hipMemcpyDeviceToHost));
+      break;
+    }
+  }
+}
+
+// Copies device data to host and checks that each element is equal to the
+// specified value
+template <typename T>
+void verifyData(T* aPtr, size_t value, MultiDData& data, allocType type, memSetType memType) {
+  auto dataH = data.height == 0 ? 1 : data.height;
+  auto dataD = data.depth == 0 ? 1 : data.depth;
+  std::unique_ptr<T[]> hostPtr = std::make_unique<T[]>(data.pitch * dataH * dataD / sizeof(T));
```

Comment:
when copying onto the host for verification I think we should be using just the extent of the data instead of the extent with pitch included.